src/c++/perf_analyzer/client_backend/openai/openai_client.cc

```diff
@@ -0,0 +1,297 @@
+
+  char* buf = reinterpret_cast<char*>(contents);
+  size_t result_bytes = size * nmemb;
+  request->response_buffer_.append(buf, result_bytes);
+  // Send response now if streaming, otherwise wait until request has been
+  // completed
+  if (request->is_stream_) {
+    auto done_signal =
+        (request->response_buffer_.find("data: [DONE]") != std::string::npos);
+    request->SendResponse(
+        done_signal /* is_final */, done_signal /* is_null */);
+  }
+
+  // ResponseHandler may be called multiple times so we overwrite
+  // RECV_END so that we always have the time of the last.
+  request->timer_.CaptureTimestamp(
+      triton::client::RequestTimers::Kind::RECV_END);
+
+  return result_bytes;
+}
+
+
+Error
+ChatCompletionClient::AsyncInfer(
+    std::function<void(InferResult*)> callback,
+    std::string& serialized_request_body, const std::string& request_id)
+{
+  if (callback == nullptr) {
+    return Error(
+        "Callback function must be provided along with AsyncInfer() call.");
+  }
+
+  auto completion_callback = [this](HttpRequest* req) {
+    auto request = static_cast<ChatCompletionRequest*>(req);
+    if (!request->is_stream_) {
+      request->SendResponse(true /* is_final */, false /* is_null */);
+    }
+    request->timer_.CaptureTimestamp(
+        triton::client::RequestTimers::Kind::REQUEST_END);
+    UpdateInferStat(request->timer_);
+  };
+  std::unique_ptr<HttpRequest> request(new ChatCompletionRequest(
+      std::move(completion_callback), std::move(callback), request_id,
+      verbose_));
+  auto raw_request = static_cast<ChatCompletionRequest*>(request.get());
+  raw_request->timer_.CaptureTimestamp(
+      triton::client::RequestTimers::Kind::REQUEST_START);
+  request->AddInput(
+      reinterpret_cast<uint8_t*>(serialized_request_body.data()),
+      serialized_request_body.size());
+
+  CURL* multi_easy_handle = curl_easy_init();
+  Error err = PreRunProcessing(multi_easy_handle, raw_request);
+  if (!err.IsOk()) {
+    curl_easy_cleanup(multi_easy_handle);
+    return err;
+  }
+
+  raw_request->timer_.CaptureTimestamp(
+      triton::client::RequestTimers::Kind::SEND_START);
+  Send(multi_easy_handle, std::move(request));
+  return Error::Success;
+}
+
+Error
+ChatCompletionClient::PreRunProcessing(
+    CURL* curl, ChatCompletionRequest* request)
+{
+  curl_easy_setopt(curl, CURLOPT_URL, url_.c_str());
+  curl_easy_setopt(curl, CURLOPT_USERAGENT, "libcurl-agent/1.0");
+  curl_easy_setopt(curl, CURLOPT_POST, 1L);
+  curl_easy_setopt(curl, CURLOPT_TCP_NODELAY, 1L);
+
+  if (verbose_) {
+    curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
+  }
+
+  const long buffer_byte_size = 16 * 1024 * 1024;
+  curl_easy_setopt(curl, CURLOPT_UPLOAD_BUFFERSIZE, buffer_byte_size);
+  curl_easy_setopt(curl, CURLOPT_BUFFERSIZE, buffer_byte_size);
+
+  // request data provided by RequestProvider()
+  curl_easy_setopt(curl, CURLOPT_READFUNCTION, RequestProvider);
+  curl_easy_setopt(curl, CURLOPT_READDATA, request);
```

Comment:
Why use `CURLOPT_READFUNCTION`/`CURLOPT_READDATA` instead of ` CURLOPT_POSTFIELDS`/`CURLOPT_POSTFIELDSIZE`?

https://curl.se/libcurl/c/CURLOPT_POST.html

Reply:
@GuanLuo

Reply:
There is no particular reason as I wasn't aware of the `CURLOPT_POSTFIELDS` flag. But I think it would be good to make the read / write symmetric, also with the read callback we can inject some logic within (i.e. capturing `SEND_END` timestamp)